### PR TITLE
Add support for multi dimension range partitioning

### PR DESCRIPTION
**Pull request by kfaraz**
## Description

This PR adds support for range partitioning on multiple dimensions. It extends on the
concept and implementation of single dimension range partitioning.

The new partition type added is `range` which corresponds to a set of Dimension Range Partition classes. `single_dim` is now treated as a `range` type partition with a single partition dimension.

The `start` and `end` values of a `DimensionRangeShardSpec` are represented
by `StringTuples`, where each String in the tuple is the value of a partition dimension.

For example, the following could be the start and end values of some shards for
partition dimensions [country, city]:

Shard Number  | Start [country, city] | End [country, city] (exclusive)
----------------|------|-----
1| null | ["France", "Paris"]
2| ["France", "Paris"] | ["UK", "London"]
3| ["UK", "London"] | null

Thus, `null < ["France", "Paris"] < ["UK", "London"] < null`
In the above shards,
- all rows with no country specified (e.g. the rows [null, null], [null, "city A"] ) would exist in shard 1
- all rows for country "Algeria" would exist in shard 1
- all rows for country "Italy" would exist in shard 2
- all rows for country "Zimbabwe" would exist in shard 3

Please note:
- `null` is used to represent both overall start and overall end (same as current behaviour for single_dim)
- `null` is used to mark the overall start/end whereas a tuple of nulls (e.g. `[null, null]`)
    can be a valid tuple of dimension values.
- After the changes here, `single_dim` would be treated as a special case of `range`. 
- The single dimension classes have been retained for backward compatibility
   but all new/existing code must use the new DimensionRange partition classes.
- Partitioning done with type `range` and a single partition dimension would still create shard specs
   of type `single_dim`

For API compatibility, the following partition specs are equivalent:

```json
{
  "type": "single_dim",
  "partitionDimension": "dim1"
}
```

```json
{
  "type": "range",
  "partitionDimensions": ["dim1"]
}
```

## Major Code Changes
 * Add `StringTuple` to represent the tuple of dimension values in an `InputRow`
 * Add `DimensionRangePartitionsSpec`, super class of `SingleDimensionPartitionsSpec`
 * Add `DimensionRangeShardSpec`, super class of `SingleDimensionShardSpec`
 * Add `DimensionRangeBucketShardSpec`, `BuildingDimensionRangeShardSpec`
 * Create instances of `DimensionRangeBucketShardSpec` when bucketing rows into partitions in
   `RangePartitionAnalysis`
 * Use `StringTuple` instead of `String` when creating partition ranges
 * In `DimensionRangeShardSpec`, check ranges of only the first dimension when pruning partitions

## Changes not in this PR
- Multi dim partitioning changes for indexing-hadoop classes

## Changes in follow up PR
- Docs changes for the new multi dim partitioning

<hr>

This PR has:
- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: core/src/main/java/org/apache/druid/indexer/partitions/MultiDimensionPartitionsSpec.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    Integer adjustedMaxPartitionSize = PartitionsSpec.resolveHistoricalNullIfNeeded(maxPartitionSize);
+
+    Property<Integer> target = Checks.checkAtMostOneNotNull(
+        DimensionBasedPartitionsSpec.TARGET_ROWS_PER_SEGMENT,
+        adjustedTargetRowsPerSegment,
+        DimensionBasedPartitionsSpec.TARGET_PARTITION_SIZE,
+        adjustedTargetPartitionSize
+    );
+
+    Property<Integer> max = Checks.checkAtMostOneNotNull(
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply core/src/main/java/org/apache/druid/indexer/partitions/MultiDimensionPartitionsSpec.java]: Removed deprecated fields from multi dim.

File: core/src/main/java/org/apache/druid/indexer/partitions/SingleDimensionPartitionsSpec.java [hunk 2]
```diff
@@ -21,83 +21,52 @@
...
-  private final Integer targetRowsPerSegment;
-  private final Integer maxRowsPerSegment;
   private final String partitionDimension;
-  private final boolean assumeGrouped;
-
-  // Values for these fields are derived from the one above:
-  private final int resolvedMaxRowPerSegment;
 
   @JsonCreator
   public SingleDimensionPartitionsSpec(
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply core/src/main/java/org/apache/druid/indexer/partitions/SingleDimensionPartitionsSpec.java:48]: Done. Thanks for the suggestion!

File: core/src/main/java/org/apache/druid/timeline/partition/MultiDimensionShardSpec.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+  public <T> PartitionChunk<T> createChunk(T obj)
+  {
+    if (numCorePartitions == UNKNOWN_NUM_CORE_PARTITIONS) {
+      return new StringPartitionChunk<>(start, end, partitionNum, obj);
+    } else {
+      return new NumberedPartitionChunk<>(partitionNum, numCorePartitions, obj);
+    }
+  }
+
+  @VisibleForTesting
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply core/src/main/java/org/apache/druid/timeline/partition/MultiDimensionShardSpec.java]: Fixed.

File: indexing-service/src/test/java/org/apache/druid/indexing/common/task/RangePartitionCachingLocalSegmentAllocatorTest.java [hunk 4]
```diff
@@ -54,12 +55,16 @@
 import java.util.Map;
 import java.util.stream.Collectors;
 
+import static org.apache.druid.timeline.partition.ShardSpecTestUtils.tupleOf;
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply indexing-service/src/test/java/org/apache/druid/indexing/common/task/RangePartitionCachingLocalSegmentAllocatorTest.java]: Replaced with `StringTuple.create()` in this test and other places.

File: core/src/main/java/org/apache/druid/indexer/partitions/MultiDimensionPartitionsSpec.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+  @JsonCreator
+  public MultiDimensionPartitionsSpec(
+      @JsonProperty(TARGET_ROWS_PER_SEGMENT) @Nullable Integer targetRowsPerSegment,
+      @JsonProperty(MAX_ROWS_PER_SEGMENT) @Nullable Integer maxRowsPerSegment,
+      @JsonProperty(PARITION_DIMENSIONS) List<String> partitionDimensions,
+      @JsonProperty("assumeGrouped") boolean assumeGrouped,  // false by default
+
+      // Deprecated properties preserved for backward compatibility:
+      @Deprecated @JsonProperty(TARGET_PARTITION_SIZE) @Nullable
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply core/src/main/java/org/apache/druid/indexer/partitions/MultiDimensionPartitionsSpec.java]: Removed.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/iterator/RangePartitionIndexTaskInputRowIteratorBuilder.java [hunk 6]
```diff
@@ -79,36 +80,46 @@ public HandlingInputRowIterator build()
...
-      );
+    for (String dimension : partitionDimensions) {
+      int dimensionValueCount = inputRow.getDimension(dimension).size();
+      if (dimensionValueCount > 1) {
+        throw new IAE(
+            "Cannot partition on multi-value dimension [%s] for input row [%s]",
+            dimension,
+            inputRow
+        );
+      } else if (valueCountPredicate.test(dimensionValueCount)) {
```
Comment (abhishekagarwal87) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/iterator/RangePartitionIndexTaskInputRowIteratorBuilder.java:117]: so the `valueCountPredicate` is `dimValueCount -> dimValueCount > 1` but if dimensionValueCount is greater than 1, we will throw an exception. so this should always return false. isn't it?

Comment (abhishekagarwal87) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/iterator/RangePartitionIndexTaskInputRowIteratorBuilder.java:117]: ok. there is one more value predicate. thinking more on it, `inputRow -> isRowHandled(inputRow, partitionDimensions, dimValueCount -> dimValueCount > 1)` can very well be inputRow -> isRowHandled(inputRow, partitionDimensions, dimValueCount -> false)`. Just trying to understand this logic better.

**Review by abhishekagarwal87 — COMMENTED** [review]
Few comments. LGTM overall.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/distribution/ArrayOfStringTuplesSerDe.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+  @Override
+  public byte[] serializeToByteArray(StringTuple[] items)
+  {
+    int length = 0;
+    final byte[][] itemsBytes = new byte[items.length][];
+    for (int i = 0; i < items.length; i++) {
+      itemsBytes[i] = STRINGS_SERDE.serializeToByteArray(items[i].toArray());
+      length += itemsBytes[i].length + Integer.BYTES;
+      length += items[i].size() + Integer.BYTES;
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/distribution/ArrayOfStringTuplesSerDe.java]: Fixed. Thanks!

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/PartialRangeSegmentGenerateTask.java [hunk 8]
```diff
@@ -92,20 +92,20 @@ public PartialRangeSegmentGenerateTask(
...
   }
 
-  private static String getPartitionDimension(ParallelIndexIngestionSpec ingestionSpec)
+  private static List<String> getPartitionDimensions(ParallelIndexIngestionSpec ingestionSpec)
   {
     PartitionsSpec partitionsSpec = ingestionSpec.getTuningConfig().getPartitionsSpec();
     Preconditions.checkArgument(
-        partitionsSpec instanceof SingleDimensionPartitionsSpec,
+        partitionsSpec instanceof MultiDimensionPartitionsSpec,
         "%s partitionsSpec required",
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/PartialRangeSegmentGenerateTask.java]: Done.

File: core/src/main/java/org/apache/druid/indexer/partitions/MultiDimensionPartitionsSpec.java [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+ * Range partition a segment by multiple dimensions.
+ */
+public class MultiDimensionPartitionsSpec implements DimensionBasedPartitionsSpec
+{
+  public static final String NAME = "multi_dim";
+
+  private final Integer targetRowsPerSegment;
+  private final Integer maxRowsPerSegment;
+  private final List<String> partitionDimensions;
+  private final boolean assumeGrouped;
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply core/src/main/java/org/apache/druid/indexer/partitions/MultiDimensionPartitionsSpec.java:51]: Added javadoc at the class level.

**Review by abhishekagarwal87 — APPROVED** [review]

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/iterator/RangePartitionIndexTaskInputRowIteratorBuilder.java (hunk 6 above)
Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/iterator/RangePartitionIndexTaskInputRowIteratorBuilder.java]: Refactored this logic to have two methods:
- `ensureNoMultiValueDimensions()` which checks for multiple values and throws IAE
- `hasEmptyDimensions()`

File: core/src/main/java/org/apache/druid/timeline/partition/MultiDimensionShardSpec.java [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+  @Override
+  public int hashCode()
+  {
+    return Objects.hash(dimensions, start, end, partitionNum, numCorePartitions);
+  }
+
+  @Override
+  public String toString()
+  {
+    return "SingleDimensionShardSpec{" +
```
Comment (gianm) [inline core/src/main/java/org/apache/druid/timeline/partition/MultiDimensionShardSpec.java:236]: Should be MultiDimensionShardSpec?

**Comment by gianm** [comment]
The overall design / API looks good based on the PR description (I haven't looked at the code). It'll be a great feature.

I wanted to ask about compatibility during rolling updates and rollbacks:

1. If someone updates to the new version, but doesn't use the new feature (i.e. they keep submitting ingest specs with `single_dim` shard specs), will that lead to segments getting written with `single_dim` or `multi_dim` shard specs? Hopefully it's `single_dim`, since that means people can safely roll back without losing the ability to read their shard specs.
2. If someone is in the middle of a rolling update to the new version, and has some data servers updated but not others, will that be safe? Or will some servers write out `multi_dim` shard specs or other `multi_dim` things that other servers on the older version cannot read?

**Comment by kfaraz** [comment]
>     1. If someone updates to the new version, but doesn't use the new feature (i.e. they keep submitting ingest specs with `single_dim` shard specs), will that lead to segments getting written with `single_dim` or `multi_dim` shard specs? Hopefully it's `single_dim`, since that means people can safely roll back without losing the ability to load their existing segments.

Thanks for pointing this out, @gianm . I had originally intended to add this logic to ensure backward compatibility but I had missed it.
I have added it now by making `SingleDimensionShardSpec` extend `MultiDimensionShardSpec`.
Thus, while converting a `BuildingMultiDimensionShardSpec`, a `SingleDimensionShardSpec` is created if there is only one dimension, otherwise a `MultiDimensionShardSpec` is created.

```java
@Override
  public MultiDimensionShardSpec convert(int numCorePartitions)
  {
    return dimensions != null && dimensions.size() == 1
           ? new SingleDimensionShardSpec(...) 
           : new MultiDimensionShardSpec(...);
  }
```

The only drawback is that if someone tried to partition their data using type `multi_dim` with a single dimension specified in the `partitionDimensions` field, the shard specs created would still have type `single_dim`. But I guess that is to be expected.

>     2. If someone is in the middle of a rolling update to the new version, and has some servers updated but not others, will that be safe? Or will some servers write out `multi_dim` shard specs or other `multi_dim` things that other servers on the older version cannot read?

The above mentioned approach will take care of this as well, because servers will __output `multi_dim` shard specs__ only if a user explicitly chooses `multi_dim` partitioning __with more than one dimensions__.

**Comment by abhishekagarwal87** [comment]
Thank you @kfaraz for your contribution.

**Comment by kfaraz** [comment]
No, @hqx871 , I haven't added support for Hadoop ingestion yet. 🙂

**Comment by hqx871** [comment]
@kfaraz thanks for your great work. This feature really helps improve our large datasource queries, as it prune most segments. Besides, I also found that the DimensionRangeShardSpec lookup may be slow when there are too many shards. I tried to sort the shardSpecs and use binary search to fast the routine. Hope this is helpful to somebody else.

**Comment by hqx871** [comment]
By lookup is slow I mean the follow code:
```
private static ShardSpecLookup createLookup(List<? extends ShardSpec> shardSpecs)
  {
    return (long timestamp, InputRow row) -> {
      for (ShardSpec spec : shardSpecs) {
        if (((DimensionRangeShardSpec) spec).isInChunk(row)) {
          return spec;
        }
      }
      throw new ISE("row[%s] doesn't fit in any shard[%s]", row, shardSpecs);
    };
  }
```

**Comment by kfaraz** [comment]
Thanks a lot for the feedback, @hqx871 ! I am really glad you find it useful. 🙂

Thanks for the suggestion on sorting the shard specs, I will look into it.

**Comment by abhishekagarwal87** [comment]
@hqx871 - you could contribute that patch.

**Comment by hqx871** [comment]
As we only use the hadoop based batch ingestion in our company and I know little about the native batch ingestion. So modify it's code is hard to me and to test as well.

**Comment by hqx871** [comment]
I have make a try to do this

**Pull request merged**